File: Lydian.Unity.Automapper/Core/TypeMappingFactory.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;

namespace Lydian.Unity.Automapper.Core
{
	internal sealed class TypeMappingFactory : ITypeMappingFactory
	{
		public IEnumerable<TypeMapping> CreateMappings(MappingBehaviors behaviors, AutomapperConfig configurationDetails, params Type[] types)
		{
			var results = from availableInterface in types.Where(type => type.IsInterface)
														  .Where(configurationDetails.IsMappable)
						  from concrete in types.Where(type => !type.IsInterface)
												.Where(configurationDetails.IsMappable)
						  from concreteInterface in concrete.GetGenericallyOpenInterfaces()
															.Where(ci => ci.Item1 == availableInterface)
						  let matchingPair = new { concrete, concreteInterface }
						  group matchingPair by matchingPair.concreteInterface.Item2 into mappingsForAnInterface
						  from mapping in mappingsForAnInterface
						  select new TypeMapping(mapping.concreteInterface.Item2, mapping.concrete);

			if (!behaviors.HasFlag(MappingBehaviors.CollectionRegistration))
				return results.ToArray();

			return CreateAcrMappings(results, configurationDetails)
							.Union(results)
							.ToArray();
		}

		private static IEnumerable<TypeMapping> CreateAcrMappings(IEnumerable<TypeMapping> mappings, AutomapperConfig configurationDetails)
		{
			Contract.Requires(mappings != null);
			return mappings
					.GroupBy(m => m.From)
					.Where(g => g.Count() > 1 || configurationDetails.IsMultimap(g.Key))
					.Select(g => new TypeMapping(GetGenericTypeSafely(typeof(IEnumerable<>), g.Key), GetGenericTypeSafely(typeof(UnityCollectionFacade<>), g.Key)));
		}

		private static Type GetGenericTypeSafely(Type source, params Type[] args)
		{
			if (!source.IsGenericTypeDefinition)
				throw new ArgumentException("source must be a generic type definition", "source");
			if (source.GetGenericArguments().Length != args.Length)
				throw new ArgumentException("incorrect number of arguments", "args");
			return source.MakeGenericType(args);
		}
	}
}

File: Lydian.Unity.Automapper.Fs/TypeMappingFactory.fs
/// Creates type mappings from a supplied set of types.
module internal Lydian.Unity.Automapper.Core.TypeMappingFactory

open Lydian.Unity.Automapper
open Microsoft.Practices.Unity
open System
open System.Collections.Generic

let private getGenericTypeSafely(destination: Type, genericParameter) = 
    if not destination.IsGenericTypeDefinition then 
        raise <| ArgumentException("source must be a generic type definition", "source")
    if destination.GetGenericArguments().Length <> 1 then 
        raise <| ArgumentException("incorrect number of arguments", "destination")
    destination.MakeGenericType [| genericParameter |]

let private getGenericallyOpenInterfaces(concrete: Type) = 
    query { 
        for concreteInterface in concrete.GetInterfaces() do
            let comparisonInterface = 
                if concreteInterface.IsGenericType then concreteInterface.GetGenericTypeDefinition()
                else concreteInterface
            
            let destinationInterface = 
                if concrete.IsGenericType then comparisonInterface
                else concreteInterface
            
            select(comparisonInterface, destinationInterface)
    }

/// Creates auto-generated mappings from a set of types from which to perform registrations on.
let private createMappingsFromTypes(behaviors: MappingBehaviors, configurationDetails: AutomapperConfigData, 
                                    types: Type seq) = 
    let results = 
        query { 
            for availableInterface in types
                                      |> Seq.filter(fun t -> t.IsInterface)
                                      |> Seq.filter configurationDetails.IsMappable do
                for concrete in types
                                |> Seq.filter(fun t -> not t.IsInterface)
                                |> Seq.filter configurationDetails.IsMappable do
                    for concreteInterface in (getGenericallyOpenInterfaces(concrete) 
                                              |> Seq.filter(fun (comp, _) -> comp = availableInterface)) do
                        select(concrete, concreteInterface)
        }
    
    let results = 
        query { 
            for result in results do
                groupBy (result
                         |> snd
                         |> snd) into mappingsForAnInterface
                for mapping in mappingsForAnInterface do
                    select((mapping
                            |> snd
                            |> snd), (fst mapping))
        }
    
    results |> Seq.toList

/// Creates Automatic Collection Registrations from a set of mappings and configuration data.
let private createAcrMappings (configurationDetails: AutomapperConfigData) mappings = 
    mappings @ (mappings
                |> Seq.countBy fst
                |> Seq.filter(fun (key, count) -> count > 1 || key |> configurationDetails.IsMultimap)
                |> Seq.map
                       (fun (key, count) -> 
                       getGenericTypeSafely(typedefof<IEnumerable<_>>, key), 
                       getGenericTypeSafely(typedefof<UnityCollectionFacade<_>>, key))
                |> Seq.toList)

/// Creates a set of type <-> type mappings using the supplied behaviors and configuration over the provided set of types.
let createMappings(behaviors: MappingBehaviors, configuration, types) = 
    let createMappings = 
        if behaviors.HasFlag(MappingBehaviors.CollectionRegistration) then 
            createMappingsFromTypes >> (createAcrMappings configuration)
        else createMappingsFromTypes
    createMappings(behaviors, configuration, types)
